### PR TITLE
Add JSON Web Key to specification

### PR DESCRIPTION
**Pull request by msporny**
This PR addresses issue #130 and issue #73 by adding the `JsonWebKey` type to the specification and the vocabulary, and ensuring that the range of `publicKeyJwk` is JSON. This value was going to be defined in the JsonWebSignature2020 specification, but since that specification has been pulled, we have to define it somewhere else and the Data Integrity specification is probably the best place for it since it also defined controller documents and verification methods.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/pull/135.html" title="Last updated on Aug 12, 2023, 7:26 PM UTC (1852514)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/135/b809599...1852514.html" title="Last updated on Aug 12, 2023, 7:26 PM UTC (1852514)">Diff</a>

### CODE REVIEW

**Review by mkhraisha — APPROVED** [review]

**Review by Wind4Greg — APPROVED** [review]

**Review by dlongley — APPROVED** [review]

File: vocab/security/vocabulary.yml
Comment (iherman) [inline vocab/security/vocabulary.yml]: Strictly speaking, this is not necessary (although it does not harm). The tool has a number default prefixes which include rdf, rdfs, dc, owl, xsd.

```suggestion
```

**Review by iherman — APPROVED** [review]
Just an editorial comment, possibly with an editorial note. It may be worth emphasizing that the `publicKeyJwk` having a `JSON` datatype means that all json terms _within_ the target are not subject to the JSON-LD context processing. Ie, the terms like `kid` or `alg` are _not_ mapped to any URL.

File: index.html
Comment (iherman) [inline index.html]:
```suggestion
  "@context": ["https://www.w3.org/ns/security/jwk/v1"],
```

File: vocab/security/vocabulary.yml
Comment (iherman) [inline vocab/security/vocabulary.yml:49]: I did not see the `<dfn>` statement in the text...

**Review by iherman — CHANGES_REQUESTED** [review]

File: vocab/security/vocabulary.yml
Comment (msporny) [reply vocab/security/vocabulary.yml]: Ah! Thank you, I checked the yml2vocab tool and didn't see the definition. Great that it exists, removing the `rdf` definition now.

File: vocab/security/vocabulary.yml
Comment (msporny) [reply vocab/security/vocabulary.yml:49]: ReSpec auto-generates it based on this line: https://github.com/w3c/vc-data-integrity/pull/135/files#diff-0eb547304658805aad788d320f10bf1f292797b5e6d745a3bf617584da017051R1094

**Review by thereisnogabe — APPROVED** [review]

File: index.html
Comment (OR13) [inline index.html]:
```suggestion
              href:     "https://www.iana.org/assignments/jose",
```

Comment (OR13) [reply index.html]: Possible this will change in the future, better to not ask for XML by file extension

File: index.html
Comment (OR13) [inline index.html]: hmmm... `privateKeyJwk` vs `secretKeyJwk`... Is the intention to make this work with MAC Signatures? and thats why you choose `secretKeyJwk` ?

File: index.html
Comment (OR13) [inline index.html]: This seems possibly not true for all secret keys.

File: vocab/security/vocabulary.yml
Comment (OR13) [inline vocab/security/vocabulary.yml:272]: Expecting to see `secretKeyJwk` defined in this vocabulary as well... otherwise how will people understand how to use it?

**Review by OR13 — CHANGES_REQUESTED** [review]
We should get more reviews on the `secretKeyJwk` part... will it be in the vocabulary? can it be used with HMAC?.

Concrete change request is to define `secretKeyJwk` in the vocabulary or remove the example.

I'm not attached to `privateKeyJwk` vs `secretKeyJwk`... perhapse it would be good to ask the experts on the JOSE and COSE mailing lists how they feel about this?

File: index.html
Comment (mprorock) [reply index.html]: this is definitely not true for all secret keys, and may yet be changing - why would we define any of this here rather than just point at rfc7517?

**Comment by mprorock** [comment]
+1 @OR13 - definitely want feedback in from the jose and cose lists

**Review by mprorock — CHANGES_REQUESTED** [review]
lets get feedback requests out to appropriate experts - clean up some naming / guidance or just point to appropriate RFCs, otherwise i think this good

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
that uses the JWK specification [[RFC7517]] to encode key types into a
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
The JSON Web Key (JWK) data model is a specific type of <a>verification method</a>
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
might be referenced in software systems; if present, the `kid` value SHOULD match the
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
The JSON Web Key data model is also capable of encoding <em>secret keys</em>, sometimes
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
              title="JSON Web Key encoding of an Ed25519 public key">
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
              title="JSON Web Key encoding of an Ed25519 secret key">
```

File: index.html
Comment (TallTed) [inline index.html]:
```suggestion
The private key example above is almost identical to the previous example of the
public key, except that the information is stored in the `secretKeyJwk` property
(rather than the `publicKeyJwk`), and the private key value is encoded in the `d`
property thereof (alongside the `x` property, which still specifies the point on
the Ed25519 curve that is associated with the public key).
```

**Review by TallTed — CHANGES_REQUESTED** [review]
Mostly small editorial stuff. One larger thing which I think is also editorial and hopefully not controversial.

**Review by dmitrizagidulin — APPROVED** [review]
:+1: overall, (agree with Mike about not defining the specific key properties like `d` in the spec, and just pointing to the RFC)

File: vocab/security/vocabulary.yml
Comment (dmitrizagidulin) [reply vocab/security/vocabulary.yml:272]: Yes, secretKeyJwk is added to the vocab in PR https://github.com/w3c/vc-data-integrity/pull/148

**Comment by OR13** [comment]
I asked the IETF lists: https://mailarchive.ietf.org/arch/msg/cose/fuTzZtvbvSF3GQFVjI-JYOyX7Lk/

I suggest we leave the PR open for another week in case they have feedback.

EDIT: also please apply @TallTed 's suggestions.

File: index.html
Comment (OR13) [reply index.html]: See https://datatracker.ietf.org/doc/draft-ietf-cose-key-thumbprint/

for concrete examples...

https://mailarchive.ietf.org/arch/msg/cose/T4Od25YoPmgOuALuuIGf1m8N4wo/

File: index.html
Comment (OR13) [reply index.html]: maybe better to refer to the spec language.... which calls this "private information"....

https://datatracker.ietf.org/doc/html/rfc7517#section-8.1

> It also records whether the  parameter conveys public or private information.

**Review by OR13 — CHANGES_REQUESTED** [review]
Please apply suggestions so that re-review is possible.

File: index.html
Comment (TallTed) [inline index.html]: Integrating [the suggestion](https://github.com/w3c/vc-data-integrity/pull/135/files#r1284562689) from @OR13 --
```suggestion
The JSON Web Key data model is also capable of encoding private information,
including <em>private keys</em> and/or <em>secret keys</em>.
```

Comment (OR13) [reply index.html]: ^ I support the language above, but in case we want to be even more precise, "private information" is not "secrete keys" or "private keys".... its parameters of "secrete keys" or "private keys"... for example:

https://www.iana.org/assignments/jose/jose.xhtml#web-key-set-parameters

Screen Shot 2023-08-04 at 5 10 31 PM

Just look at all those RSA things that are not "d" ... that are private... that is why I tried to correct the language here: 

- https://github.com/w3c/vc-data-integrity/pull/99

The advice given in data integrity regarding "d" does not actually protect people... since RSA does not use "d"... and lots of people still use RSA saddly.

Comment (TallTed) [reply index.html]: Updated language. 

> `The JSON Web Key data model is also capable of encoding private information,
including <em>private keys</em> and/or <em>secret keys</em>.`

Comment (dlongley) [reply index.html]: > I support the language above, but in case we want to be even more precise, "private information" is not "secrete keys" or "private keys".... its parameters of "secrete keys" or "private keys"... for example:

Yes, and I believe this is one of the challenges with some key formats. I think it's better for app developers to be able to treat key values as opaque and completely omit secret keys without having to know the details around validating and parsing (potentially) secret keys to look for specific internal components, etc. This is best left to another layer.

Comment (dlongley) [reply index.html]: > The advice given in data integrity regarding "d" does not actually protect people... since RSA does not use "d"... and lots of people still use RSA saddly.

This might have been a typo; RSA does use "d" -- that's where it came from originally:

https://en.wikipedia.org/wiki/RSA_(cryptosystem)#Operation

Comment (OR13) [reply index.html]: Yes, I meant to say... omitting ONLY `d` (which is the current guidance) is bad advice... 

Just like it would be bad advice to say that only uranium is radioactive... There are other radioactive elements... and there are other "private information"  parameters that should not be seen in public keys.

File: index.html
Comment (selfissued) [reply index.html]: I agree that privateKeyJwk is a better name than secretKeyJwk.

File: index.html
Comment (dlongley) [reply index.html]: > Yes, I meant to say... omitting ONLY `d` (which is the current guidance) is bad advice...

Agreed -- it's good as an example but that example should say that it's insufficient to just omit that one thing and that `<reference X>` must be checked and every possible private or secret (whichever terminology is appropriate) component listed similarly omitted.

**Comment by dlongley** [comment]
> I'm not attached to `privateKeyJwk` vs `secretKeyJwk`... perhaps it would be good to ask the experts on the JOSE and COSE mailing lists how they feel about this?

Sure, but note that we're defining the interface / expression here where the most important audience is app developers who should be thoroughly made aware of "secret" information so it can be dealt with appropriately. Whether it's the best name in the opinion of a particular cryptographer or a perfect fit for a particular cryptographic scheme is not right the measure for the name, IMO. It doesn't really matter what the underlying crypto is; the value it holds is "secret" and the type for the value is a JWK -- any JWK. That's where the JOSE (or COSE if it were `secretKeyCwk`) layer would come in -- and we should just reuse those specs for the values of the properties.

File: index.html
Comment (msporny) [reply index.html]: > Is the intention to make this work with MAC Signatures? and thats why you choose secretKeyJwk ?

Yes, that's one of the reasons.

The other reason is due to the English definition of the word "private" vs. the English definition of the word "secret":

private: intended for or restricted to the use of a particular person, group, or class
secret: kept from knowledge or view; something kept hidden or unexplained

IOW, the word "secret", that is -- "do not share this with others", is closer to the intended usage of the property than "private", which can be misinterpreted as "its ok to share this information with a group" (which is almost never the case with a symmetric or a private key).

File: index.html
Comment (msporny) [reply index.html]: > This seems possibly not true for all secret keys.

The language you are pointing to is providing a description of a specific example. It is not claiming that the example is true for all secret keys.

> this is definitely not true for all secret keys, and may yet be changing - why would we define any of this here rather than just point at rfc7517?

Please re-read the /description of the example/. We are not defining anything here, merely explaining what an example is expressing.

**Comment by msporny** [comment]
> We should get more reviews on the secretKeyJwk part... will it be in the vocabulary? can it be used with HMAC?.

Yes, it will be in the vocabulary. Yes, it can be used w/ HMAC (and anything else JWK can do).

> Concrete change request is to define secretKeyJwk in the vocabulary or remove the example.

That is being done in PR https://github.com/w3c/vc-data-integrity/pull/148.

> I'm not attached to privateKeyJwk vs secretKeyJwk... perhapse it would be good to ask the experts on the JOSE and COSE mailing lists how they feel about this?

I see that you've sent an email to those mailing lists. I will hold this PR an additional week for feedback and then merge. If we get feedback at any point beyond that, we can raise another PR to apply the feedback.

> Please apply suggestions so that re-review is possible.

Done.

@mprorock wrote:
> just point to appropriate RFCs
@dmitrizagidulin wrote:
> agree with Mike about not defining the specific key properties like d in the spec, and just pointing to the RFC

Can you guys point to where we're defining things that are in the RFC in this specification? We /do/ point to the RFC with this language (and DO NOT re-define properties in this spec and/or PR):

> The JSON Web Key (JWK) data model is a specific type of <a>verification method</a> that uses the JWK specification [[RFC7517]] to encode key types into a set of parameters. To encode a JSON Web Key, the <a>verification method</a> `type` MUST be set to `JsonWebKey` and the `publicKeyJwk` or `secretKeyJwk` value MUST be a valid JWK value as described in [[RFC7517]].

What am I missing?

Requesting a re-review from each of you.

File: vocab/security/vocabulary.yml
Comment (iherman) [reply vocab/security/vocabulary.yml:49]: That is correct for the class specification. However, #148, that adds both the `secretKeyJwk` and `publicKeyJwk` properties rely on the presence of `<dfn`> for those. I see it for `publicKeyJwk` at https://pr-preview.s3.amazonaws.com/w3c/vc-data-integrity/pull/135.html#dfn-publickeyjwk, I did not find it for `secretKeyJwk` :-(

**Review by iherman — CHANGES_REQUESTED** [review]

File: vocab/security/vocabulary.yml
Comment (msporny) [reply vocab/security/vocabulary.yml:49]: Ah, right. I'll raise another PR to refactor the sections and put the definitions where they belong (in the sections defining both the `JsonWebKey` type and the `Multikey` type.

Comment (msporny) [reply vocab/security/vocabulary.yml:49]: PR #152 has been raised to address this issue.

File: vocab/security/vocabulary.yml
Comment (msporny) [reply vocab/security/vocabulary.yml:272]: Definitions added in PR #152

**Review by TallTed — APPROVED** [review]

**Comment by OR13** [comment]
I agree with @iherman  that `secretKeyJwk` or `privateKeyJwk` needs to be defined for this to be merged... and the same applies to `publicKeyMultibase` or `secretKeyMultibase` or `privateKeyMultibase`.

See also https://github.com/w3c/vc-data-integrity/pull/150/files#diff-af0ab1b57396186235708fdf500251dfc3f9b18aeeb992f927979d9630889c28R21

and https://github.com/w3c/vc-data-integrity/pull/149/files#diff-be8086ba9b3e09e91dd045c04051b6e1d0d2a10105db9d380835cb1151fd5470R29

Seems blocked pending https://github.com/w3c/vc-data-integrity/pull/152#pullrequestreview-1568433288

**Review by OR13 — APPROVED** [review]
Pending merge / approval of https://github.com/w3c/vc-data-integrity/pull/152#pullrequestreview-1568433288

File: index.html
Comment (iherman) [inline index.html:1163]: Why do we need yet another context here? The definition of `publicKeyJwk` stipulates that its object is of datatype JSON, meaning that the properties within that object are ignored by any JSON-LD processing. If so, then I do not see any reason using a different context (which I have not seen defined anywhere).

Proposal: just use the usual security context here.

File: index.html
Comment (iherman) [inline index.html:1202]: Same comment as above. Except that we use here `secretKeyJwk`, but I presume the range of that property is also JSON

**Review by iherman — APPROVED** [review]
I have made some inline comments.

I must say I am lost among all the PRs that refer to various property names, anchors, ranges, etc. I am of the opinion at this point that all these PRs should be merged asap to allow for a comprehensive review of the vocabulary, domains, ranges, etc, possibly leading to new PRs. The current situation is simply too chaotic in my view.

File: index.html
Comment (OR13) [reply index.html:1163]: We don't need another context here.

I will object if one is required... however, I think its fine to ship many smaller contexts from `w3id.org` (which is not the W3C)... IMO `W3C` examples should use `W3C` context URLs.

Comment (OR13) [reply index.html:1163]: Also, seems we are following a different style here than the core context, where we seem to prefer bundling, that inconsistency, feels... not good to me.

Comment (dlongley) [reply index.html:1163]: @OR13,

The `@vocab` in the core v2 context is what is driving the need to bundle. Whenever that isn't used, bundling should be avoided (unless there's some other very good reason to do it). That's the consistent rule we're applying.

Comment (msporny) [reply index.html:1163]: Raised https://github.com/w3c/vc-data-integrity/issues/164 to track @OR13's concern, which seems to be more of a misunderstanding than a misalignment.

File: index.html
Comment (msporny) [reply index.html:1202]: Same response as above, yes, range is JSON.

**Comment by msporny** [comment]
PR #152 has been merged into this PR. It has been two weeks since we asked for feedback from the COSE mailing list (and no feedback has been given). If we get feedback in the future, we can always raise an issue and modify this via a future PR. In the meantime, this PR needs to go in so we have the appropriate anchors to finish up the vocabulary and JSON Web Key sections of the specification.

**Comment by msporny** [comment]
Normative, multiple reviews, changes requested and made, no objections, merging.

**Pull request merged**
